library/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/processor/SnapshotProcessorTest.kt

```diff
@@ -0,0 +1,463 @@
+    }
+
+    @Test
+    fun `M flatten the tree using BFS W process { snapshot n levels }`(forge: Forge) {
+        // Given
+        val requiredLevels = forge.anInt(min = 1, max = 9)
+        val fakeSnapshot = forge.aSnapshot(requiredLevels)
+
+        // When
+        testedProcessor.process(fakeSnapshot)
+
+        // Then
+        val stack = Stack<Node>()
+        val expectedList = LinkedList<MobileSegment.Wireframe>()
+        stack.push(fakeSnapshot)
+        while (stack.isNotEmpty()) {
+            val node = stack.pop()
+            expectedList.addAll(node!!.wireframes)
+            for (i in node.children.count() - 1 downTo 0) {
+                stack.push(node.children[i])
+            }
+        }
+        val captor = argumentCaptor<EnrichedRecord>()
+        verify(mockWriter).write(captor.capture())
+        val fullSnapshotRecord = captor.firstValue.records[1] as
+            MobileSegment.MobileRecord.MobileFullSnapshotRecord
+        assertThat(fullSnapshotRecord.timestamp).isEqualTo(fakeTimestamp)
+        assertThat(fullSnapshotRecord.data.wireframes).isEqualTo(expectedList)
+    }
+
+    @Test
+    fun `M filter out the completely covered wireframes W process`(forge: Forge) {
+        // Given
+        var fakeSnapshot = forge.aSnapshot(2)
+        val topWireframeId = forge.aLong(min = 0)
+        val topWireframe = fakeSnapshot.children[1].wireframes[0].copy(topWireframeId)
+        val childrenWithCoverUpWireframe = fakeSnapshot.children + Node(
+            wireframes =
+            listOf(topWireframe)
+        )
+        fakeSnapshot = fakeSnapshot.copy(children = childrenWithCoverUpWireframe)
+
+        // When
+        testedProcessor.process(fakeSnapshot)
+
+        // Then
+        // The added wireframe completely covers the previous one + its children in the list. The
+        // previous wireframe and it children will therefore be removed.
+        val expectedList = fakeSnapshot.wireframes +
+            fakeSnapshot.children[0].wireframes +
+            fakeSnapshot.children[0].children[0].wireframes +
+            fakeSnapshot.children[0].children[1].wireframes +
+            topWireframe
+        val captor = argumentCaptor<EnrichedRecord>()
+        verify(mockWriter).write(captor.capture())
+        val fullSnapshotRecord = captor.firstValue.records[1] as
+            MobileSegment.MobileRecord.MobileFullSnapshotRecord
+        assertThat(fullSnapshotRecord.timestamp).isEqualTo(fakeTimestamp)
+        assertThat(fullSnapshotRecord.data.wireframes).isEqualTo(expectedList)
+    }
+
+    @Test
+    fun `M filter out wireframes with invalid width W process`(forge: Forge) {
+        // Given
+        var fakeSnapshot = forge.aSnapshot(1)
+        val invalidWidthWireframe: MobileSegment.Wireframe = if (forge.aBool()) {
+            forge.getForgery<MobileSegment.Wireframe.ShapeWireframe>()
+                .copy(width = forge.aLong(max = 0))
```

Comment:
`max` is exclusive, so if we want to cover cases where `width == 0` as in the code, we need to use `max = 1`. Same for the `else` block below.